include/tvm/relay/vm/vm.h

```diff
@@ -0,0 +1,284 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file tvm/relay/vm/vm.h
+ * \brief A virtual machine for executing Relay programs.
+ */
+#ifndef TVM_RELAY_VM_VM_H_
+#define TVM_RELAY_VM_VM_H_
+
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/logging.h>
+#include <tvm/runtime/memory_manager.h>
+#include <tvm/runtime/object.h>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include <memory>
+
+namespace tvm {
+namespace relay {
+namespace vm {
+
+using namespace tvm::runtime;
+
+/*! \brief A register name. */
+using RegName = size_t;
+
+/*! \brief An enumeration of Relay's opcodes.
+ *
+ * The opcode is used to implement instruction
+ * as a tagged union.
+*/
+enum struct Opcode {
+  Move,
+  Ret,
+  Invoke,
+  InvokeClosure,
+  InvokePacked,
+  AllocTensor,
+  AllocDatatype,
+  AllocClosure,
+  GetField,
+  If,
+  Select,
+  LoadConst,
+  Goto
+};
+
+
+struct Instruction {
+  Opcode op;
+
+  // Destination register that the opcode writes to
+  RegName dst;
+
+  union {
+    struct /* AllocTensorOperands */ {
+      RegName shape_register;
+      DLDataType dtype;
+    };
+    struct /* InvokeClosureOperands */ {
+      RegName closure;
+      size_t closure_args_num;
+      RegName* closure_args;
+    };
+    struct /* ReturnOperands */ {
+      RegName result;
+    };
+    struct /* MoveOperands */ {
+      RegName from;
+    };
+    struct /* PackedOperands */ {
+      size_t packed_index;
```

Comment:
avoid use size_t for instruction fields(64/32 bit platform incompatible problems) Use int64_t